torch/autograd/functional.py

```diff
@@ -0,0 +1,717 @@
+                    raise RuntimeError("The jacobian of the user-provided function is independent of "
+                                       "input {}. This is not allowed in strict mode when create_graph=True.".format(i))
+                else:
+                    raise RuntimeError("The hessian of the user-provided function is independent of "
+                                       "input {}. This is not allowed in strict mode when create_graph=True.".format(i))
+
+        res += (grads_i,)
+
+    return res
+
+# Public API
+
+def vjp(func, inputs, v=None, create_graph=False, strict=False):
+    r"""Function that computes the dot product between a vector ``v`` and the Jacobian of
+    the given function at the point given by the inputs.
+
+    Args:
+        func (function): a Python function that takes Tensor inputs and returns
+            a tuple of Tensor or a Tensor.
+        inputs (tuple of Tensor or Tensor): inputs to the function ``func``.
+        v (tuple of Tensor or Tensor): The vector that will multiply the Jacobian. Must be the
+            same size as the output of ``func``. This argument is optional when
+            ``func``'s output contains a single element and (if it is not provided) will be set as a Tensor
+            containing a single ``1``.
+        create_graph (bool, optional): If ``True``, both the output and result will be
+            computed in a differentiable way. Note that when strict is ``False``, the result can not
+            require gradients or be disconnected from the inputs.
+            Defaults to ``False``.
+        strict (bool, optional): If ``True``, an error will be raised when we detect that there exists an input
+            such that all the outputs are independent of it. If ``False``, we return zeros as the vjp for
+            said inputs, which is the expected mathematical value.
+            Defaults to ``False``.
+
+    Returns:
+        output (tuple of Tensor or Tensor): output of ``func(inputs)``
+        result (tuple of Tensor or Tensor): result of the dot product with the same shape
+            as the inputs.
+
+    Example::
+
+        >>> def exp_reducer(x):
+        ...   return x.exp().sum(dim=1)
+        >>> inputs = torch.rand(4, 4)
+        >>> v = torch.ones(4)
+        >>> vjp(exp_reducer, inputs, v)
+        (tensor([5.7817, 7.2458, 5.7830, 6.7782]),
+         tensor([[1.4458, 1.3962, 1.3042, 1.6354],
+                [2.1288, 1.0652, 1.5483, 2.5035],
+                [2.2046, 1.1292, 1.1432, 1.3059],
+                [1.3225, 1.6652, 1.7753, 2.0152]]))
+
+        >>> vjp(exp_reducer, inputs, v, create_graph=True)
+        (tensor([5.7817, 7.2458, 5.7830, 6.7782], grad_fn=<SumBackward1>),
+         tensor([[1.4458, 1.3962, 1.3042, 1.6354],
+                [2.1288, 1.0652, 1.5483, 2.5035],
+                [2.2046, 1.1292, 1.1432, 1.3059],
+                [1.3225, 1.6652, 1.7753, 2.0152]], grad_fn=<MulBackward0>))
+
+        >>> def adder(x, y):
```

Comment:
maybe it makes sense to have an example of multiple inputs and multiple outputs?

Reply:
My problem is that it is hard to have an example that is big enough so that the dimensions are readable and not just many brackets that are hard to read. While being small enough not to spam the screen too much.
This is why I stopped at multiple inputs only.

Reply:
that sounds fine then.